nixos/github-runner: add support for ephemeral and PAT registrations

veehaitch opened the pull request:
###### Description of changes

Although [neither](https://docs.github.com/en/actions/hosting-your-own-runners/adding-self-hosted-runners) [documented](https://docs.github.com/en/actions/hosting-your-own-runners/autoscaling-with-self-hosted-runners#authentication-requirements) nor suggested anywhere, the `config.sh` script has a `--pat` option which one can use for new runner registrations instead of `--token`. The help of the `config.sh` script does not suggest one could use it for runner registration: 

```
--pat GitHub personal access token used for checking network connectivity when executing `./run.sh --check`
```

Yet, [it works](https://github.com/actions/runner/blob/052ac521b0ee673dbacf79e71d08e2224a1fd3a5/src/Runner.Listener/Configuration/ConfigurationManager.cs#L555). Supersedes #181823.

Moreover, adds support for ephemeral runners which de-register and wipe their state after processing one job. If the `ephemeral` option is set to `true`, sets the unit's `Restart=` option to `on-success`. Using this option makes mostly sense if using a PAT as `tokenFile`.

Also fixes a bug which uses the runtime directory instead of the state directory as a `RUNNER_ROOT` which breaks the path of some files.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

veehaitch committed:
nixos/github-runner: use state instead of runtime dir as `RUNNER_ROOT`

Using the runtime directory as `RUNNER_ROOT` is wrong. We should always
use the state directory like we already do when invoking the runner
configure script. Otherwise, the runner constructs the wrong path for
some files (.credentials, .runner, ...).

veehaitch committed:
nixos/github-runner: add PAT support

This commit introduces support for runner registrations through a
personal access token (PAT). To use a PAT instead of a registration
token, place an appropriately scoped PAT in `tokenFile`. If the file
contains a PAT, the configuration script queries a new runner
registration token. Using a runner registration token directly continues
to work as before.

veehaitch committed:
nixos/github-runner: add support for ephemeral runners

Adds the module option `ephemeral`. If set to true, configures the
runner registration with the `--ephemeral` option. This causes the
runner to exit after processing a single job, to de-register itself, and
to delete its configuration. Afterward, systemd restarts the service
which triggers a new ephemeral registration with a clean state.

veehaitch committed:
release-notes: add github-runner support for PAT and ephemeral

newAM reviewed (changes requested):
I tried this with an existing runner + registration token and got an error upon switching:

```text
Jul 21 07:49:28 hostname l42ych29m0nprh8m9c6fjnhxpc6m18n4-github-runner-unconfigure.sh[390651]: Config has changed, removing old runner state.
Jul 21 07:49:28 hostname l42ych29m0nprh8m9c6fjnhxpc6m18n4-github-runner-unconfigure.sh[390651]: The old runner will still appear in the GitHub Actions UI. You have to remove it manually.
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390694]: Configuring GitHub Actions Runner
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390750]: touch: cannot touch '.env': Read-only file system
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390706]: ./env.sh: line 37: .path: Read-only file system
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390706]: ./env.sh: line 32: .env: Read-only file system
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390706]: ./env.sh: line 32: .env: Read-only file system
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: --------------------------------------------------------------------------------
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |        ____ _ _   _   _       _          _        _   _                      |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |       / ___(_) |_| | | |_   _| |__      / \   ___| |_(_) ___  _ __  ___      |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |      | |  _| | __| |_| | | | | '_ \    / _ \ / __| __| |/ _ \| '_ \/ __|     |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |      | |_| | | |_|  _  | |_| | |_) |  / ___ \ (__| |_| | (_) | | | \__ \     |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |       \____|_|\__|_| |_|\__,_|_.__/  /_/   \_\___|\__|_|\___/|_| |_|___/     |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |                                                                              |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |                       Self-hosted runner registration                        |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: |                                                                              |
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: --------------------------------------------------------------------------------
Jul 21 07:49:28 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: # Authentication
Jul 21 07:49:29 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: Http response code: NotFound from 'POST https://api.github.com/actions/runner-registration'
Jul 21 07:49:29 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: {"message":"Not Found","documentation_url":"https://docs.github.com/rest"}
Jul 21 07:49:29 hostname 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: Response status code does not indicate success: 404 (Not Found).
Jul 21 07:49:29 e31230v5 8zgj59zz8k5ch3ary39kbsx3yj7jx3bh-github-runner-configure.sh[390753]: Response status code does not indicate success: 404 (Not Found).
```

`services.github-runner.replace = true;` is set, so this _should_ continue to work.

veehaitch commented:
> `services.github-runner.replace = true;` is set, so this _should_ continue to work.

Thanks for testing! Yes, this should continue to work. Did you just set `services.github-runner.ephemeral = true`?

newAM commented:
> Thanks for testing! Yes, this should continue to work. Did you just set `services.github-runner.ephemeral = true`?

This was not set.  Tried it again with a fresh registration token and it works now :thinking: 
Not reproducible so I'm happy!

veehaitch committed:
nixos/github-runner: start `Runner.Listener` directly in `ExecStart=`

The `runsvc.sh` script wraps a JavaScript script which starts
`Runner.Listener` and also handles failures. This has the downside that
the service _always_ exits with status code 0, i.e., success. This
causes frequent service restarts when running in ephemeral mode with a
faulty config as Systemd always sees a success exit status. To prevent
this, this commit changes the service config to call `Runner.Listener`
directly. The JavaScript wrapper stops the process with a SIGINT, hence,
the Systemd unit now sends a SIGINT to stop the service.

veehaitch committed:
nixos/github-runner: fix Markdown docs

veehaitch commented:
Rebased to resolve conflicts.

winterqt commented on nixos/modules/services/continuous-integration/github-runner.nix:
```diff
@@ -118,6 +122,26 @@ in
+
+    ephemeral = mkOption {
+      type = types.bool;
+      description = ''
+        If enabled, causes the following behavior:
+
+        <itemizedlist>
+          <listitem><para>Passes the <literal>--ephemeral</literal> flag to the runner configuration script</para></listitem>
+          <listitem><para>De-registers and stops the runner with GitHub after it has processed one job</para></listitem>
+          <listitem><para>On stop, systemd wipes the runtime directory (this always happens, even without using the ephemeral option)</para></listitem>
+          <listitem><para>Restarts the service after its successful exit</para></listitem>
+          <listitem><para>On start, wipes the state directory and configures a new runner</para></listitem>
+        </itemizedlist>
+
+        You should only enable this option if <literal>tokenFile</literal> points to a file which contains a
+        personal access token (PAT). If you're using the option with a registration token, restarting the
+        service will fail as soon as the registration token expired.
+      '';
```
Just curious: is there a reason you didn't use Markdown for this description?

veehaitch replied:
Done!

winterqt commented on nixos/modules/services/continuous-integration/github-runner.nix:
```diff
@@ -213,17 +244,28 @@ in
                 echo "Configuring GitHub Actions Runner"
 
-                token=$(< "$STATE_DIRECTORY"/${newConfigTokenFilename})
-                RUNNER_ROOT="$STATE_DIRECTORY" ${cfg.package}/bin/config.sh \
```
Was the mismatch between this command being run and the one set in `environment` were causing issues? (Seems odd that it would ever work in the first place if that's the case, but hey, edge cases!)

As the original author of this module, why did you mismatch these two in the first place? Was it a typo, or was there some underlying reason?

veehaitch replied:
This was an error from the start. I cannot really recall what was my reasoning then but I probably tried to keep as much data as possible in the runtime directory (which Systemd removes on service stop). It didn't matter until introducing ephemeral mode as the runner didn't use the `RUNNER_ROOT` path to address files during runtime.

winterqt reviewed (commented):
I just have some comments so that I can further understand the changes, if you don't mind answering them. Thanks!

winterqt commented:
Would it be beneficial to backport at least def43d9f925c894c427bd636adb3a65223c9eef6 to 22.05?

veehaitch commented:
> Would it be beneficial to backport at least [def43d9](https://github.com/NixOS/nixpkgs/commit/def43d9f925c894c427bd636adb3a65223c9eef6) to 22.05?

I'd say yes.
